### DISCUSSION TITLE
Getting the size of the vec with file length

### DISCUSSION THREAD

**Discussion by phuongtran7** [jam1garner/binrw#163] [Q&A] [answered]
Hi,

I recently just discovered this library and it looks really nice. Thank you so much for it.

I'm trying to parse X-Plane [DSF](https://developer.x-plane.com/article/dsf-file-format-specification/) file with the library and having some problem. Here is my current solution:

```
#[binread]
#[br(magic = b"XPLNEDSF")]
struct DSF {
    version: i32,
    //#[br(count = ???)]
    //atoms: Vec<Atom>,
    #[br(seek_before = SeekFrom::End(-16), count = 16)]
    checksum: Vec<u8>,
}

#[binread]
struct Atom {
    id: i32,
    size: u32,
    #[br(count = size)]
    data: Vec<u8>,
}
```

The gist of the file format is that every file should start with 8 byte XPLNEDSF, then 4 bytes version and after that a variable length of atom, and each atom has an enum ID and the size then the data. Then the last 16 bytes are the checksum.

The problem is the file size varies, therefore the size of all the atom part varies. I'm struggling to try and set the count of the `atoms` member. Is there anyway I can infer that the file size? Like using `atoms` as `Vec<u8>` and set the count of it by using the file size minus (8 + 16) bytes?

Thank you so much for your time and help.

**Comment by phuongtran7** [comment]
I think I can parse the `Vec<Atom>` part by using `parse_with` and a custom parser. Like this:
```
struct DSF {
    version: i32,
    #[br(parse_with = custom_parser)]
    atoms: Vec<Atom>,
    #[br(seek_before = SeekFrom::End(-16), count = 16)]
    crc: Vec<u8>,
}

#[binread]
struct Atom {
    id: i32,
    size: u32,
    #[br(count = size)]
    data: Vec<u8>,
}

fn custom_parser<R: Read + Seek>(reader: &mut R, ro: &ReadOptions, _: ()) -> BinResult<Vec<Atom>> {
    let mut result = Vec::<Atom>::new();

    let tem = Atom::read_ne(reader).unwrap();
    result.push(tem);
    Ok(result)
}
```

However, I'm still not sure how to loop the custom parser until the last 16 bytes of the file yet.

**Comment by phuongtran7** [comment, marked as answer]
It seems that the solution is easier than I thought. Just seek to where I want and then save it. So everything is working now. The solution is below for anyone who comes across this in the future.

```
fn custom_parser<R: Read + Seek>(reader: &mut R, _: &ReadOptions, _: ()) -> BinResult<Vec<Atom>> {
    let mut result = Vec::<Atom>::new();
    println!("Parser is called. Length: {}", result.len());

    let mut current = reader.stream_position().unwrap();
    let last_pos = reader.seek(SeekFrom::End(-16)).unwrap();
    reader.seek(SeekFrom::Start(current)).unwrap();

    while current != last_pos {
        println!("Current position before read is: {}", current);
        let tem = Atom::read_ne(reader).unwrap();
        println!("Atom ID: {}", tem.id);
        result.push(tem);
        current = reader.stream_position().unwrap();
        println!("Current position after read is: {}", current);
    }
    Ok(result)
}
```

**Comment by csnover** [comment]
For what it’s worth, I wrote a parser for a similar atom-based file format that might be helpful to adapt for constructing objects from nested atoms: [parser](https://github.com/csnover/earthquake-rust/blob/b4a11025e5b83281005b3fc89efbe49ebcab9e61/libmactoolbox/src/quicktime/atom_parser.rs) and [usage example](https://github.com/csnover/earthquake-rust/blob/b4a11025e5b83281005b3fc89efbe49ebcab9e61/libmactoolbox/src/quicktime/vr/movie.rs#L268-L329). You will still need to do what you are doing to read the footer separately since they unfortunately didn’t go all the way to just make the hash an atom too. :-)

**Reply by phuongtran7** [reply]
Thank you so much for the info. I didn't know that the atom is used in others too. Will definitely take a look but I'm not familiar with Rust generic code yet. So it will take a while.
